src/devices/GoPiGo3/GoPiGo3.cs

```diff
@@ -0,0 +1,900 @@
+                case GrooveSensorType.None:
+                case GrooveSensorType.Custom:
+                default:
+                    return new byte[] { (byte)SpiRead8(message_type) };
+            }
+        }
+
+        /// <summary>
+        /// Get a grove input pin state
+        /// </summary>
+        /// <param name="port">The Groove Pin, one at the time Groove1Pin1, Groove1Pin2, Groove2Pin1 or Groove2Pin2</param>
+        /// /// /// <returns>Returns the pin state</returns>
+        public byte GetGrooveState(GroovePort port)
+        {
+            SpiMessageType message_type;
+            if (port == GroovePort.Groove1Pin1)
+                message_type = SpiMessageType.GetGroove1Pin1State;
+            else if (port == GroovePort.Groove1Pin2)
+                message_type = SpiMessageType.GetGroove1Pin2State;
+            else if (port == GroovePort.Groove2Pin1)
+                message_type = SpiMessageType.GetGroove2Pin1State;
+            else if (port == GroovePort.Groove2Pin2)
+                message_type = SpiMessageType.GetGroove2Pin2State;
+            else
+                throw new ArgumentException($"{nameof(GetGrooveState)} error: Pin(s) unsupported. Must get one at a time.");
+
+            byte[] outArray = { SpiAddress, (byte)message_type, 0, 0, 0, 0 };
+            var reply = SpiTransferArray(outArray);
+            if (reply[3] == 0xA5)
+                if (reply[4] == (byte)GrooveSensorState.ValidData)
+                    return reply[5];
+                else
+                    throw new IOException($"{nameof(GetGrooveState)} error: Invalid value");
+            else
+                throw new IOException($"{nameof(GetGrooveState)} error: Grove type mismatch");
+        }
+
+        /// <summary>
+        /// Get a grove input pin analog voltage
+        /// </summary>
+        /// <param name="port">The Groove Pin, one at the time Groove1Pin1, Groove1Pin2, Groove2Pin1 or Groove2Pin2</param>
+        /// /// <returns>Returns the voltage in V</returns>
+        public double GetGrooveVoltage(GroovePort port)
+        {
+            SpiMessageType message_type;
+            if (port == GroovePort.Groove1Pin1)
+                message_type = SpiMessageType.GetGroove1Pin1Voltage;
+            else if (port == GroovePort.Groove1Pin2)
+                message_type = SpiMessageType.GetGroove1Pin2Voltage;
+            else if (port == GroovePort.Groove2Pin1)
+                message_type = SpiMessageType.GetGroove2Pin1Voltage;
+            else if (port == GroovePort.Groove2Pin2)
+                message_type = SpiMessageType.GetGroove2Pin2Voltage;
+            else
+                throw new ArgumentException($"{nameof(GetGrooveVoltage)} error: Pin(s) unsupported. Must get one at a time.");
+
+            byte[] outArray = { SpiAddress, (byte)message_type, 0, 0, 0, 0, 0 };
+            var reply = SpiTransferArray(outArray);
+            if (reply[3] == 0xA5)
```

Comment:
this should use braces since it's not immediately obvious order of execution is correct